### PR TITLE
[Merged by Bors] - activation api servce: add Highest RPC for max height atx

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
part of #4614

depend on https://github.com/spacemeshos/api/pull/248

## Changes
add ActivationService::Highest() that returns the max height atx id in the node.

### CODE REVIEW

File: api/grpcserver/activation_service.go [hunk 1]
```diff
@@ -30,11 +32,24 @@ func (s *activationService) RegisterService(server *Server) {
 
 // Get implements v1.ActivationServiceServer.
 func (s *activationService) Get(ctx context.Context, request *pb.GetRequest) (*pb.GetResponse, error) {
-	if l := len(request.Id); l != types.ATXIDSize {
+	var atxId types.ATXID
```
Comment (dshulyak) [inline api/grpcserver/activation_service.go:35]: i don't see why we should have such hacks

File: api/grpcserver/activation_service.go [hunk 2]
```diff
@@ -30,11 +32,24 @@ func (s *activationService) RegisterService(server *Server) {
...
+	l := len(request.Id)
+	if l == 0 {
```
Comment (dshulyak) [inline api/grpcserver/activation_service.go]: i don't see why we should have such hacks. it is very easy to extend protobufs with one more method

Comment (fasmat) [reply api/grpcserver/activation_service.go]: `pb.GetRequest` is a generated type. I'm not sure its easy to add methods to it.

EDIT: nvm, I think I misunderstood what you meant - I assume you want an additional method besides `Get` that doesn't receive an ID? If yes, I agree 🙂

Comment (dshulyak) [reply api/grpcserver/activation_service.go]: yes method for the service :) Highest() or Commitment()

Comment (countvonzero) [reply api/grpcserver/activation_service.go]: thanks. added PR in api repository https://github.com/spacemeshos/api/pull/248
and updated the code here as well.

File: api/grpcserver/activation_service.go [hunk 3]
```diff
@@ -53,3 +56,26 @@ func (s *activationService) Get(ctx context.Context, request *pb.GetRequest) (*p
...
+				Id: &pb.ActivationId{Id: s.goldenAtx.Bytes()},
+			},
+		}, nil
+	}
+	atx, err := s.atxProvider.GetFullAtx(highest)
+	if err != nil || atx == nil {
+		return nil, status.Error(codes.NotFound, fmt.Sprintf("atx id %v not found: %v", highest, err.Error()))
+	}
+
+	if highest != atx.ID() {
```
Comment (fasmat) [inline api/grpcserver/activation_service.go]: I think this is probably not needed. If that check fails that would mean that `GetFullAtx(atxID)` returned an ATX that doesn't have `atxID` as its ID.

Comment (countvonzero) [reply api/grpcserver/activation_service.go]: i don't fully understand. it's unlikely but better not return something that's wrong in API.
why do you think it's not needed?

Comment (dshulyak) [reply api/grpcserver/activation_service.go]: i think as well that this check looks strange, it is a huge programmer mistake if it happens.
there should be a way not to write things like this

**Review by dshulyak — APPROVED** [review]

File: api/grpcserver/activation_service.go (hunk 3 above)
Comment (fasmat) [reply api/grpcserver/activation_service.go]: @countvonzero: why should a query to db for an ATX with a specific ID return an ATX with a different ID and no error? I hope we have tests in place that this cannot happen.

It's comparable to:

```golang
f, err := os.Open("test.txt")
if err != nil {
    return err
}
if f.Name() != "test.txt" {
    return fmt.Errorf("wrong file name")
}
```

I wouldn't think the second `if` is necessary.

Comment (countvonzero) [reply api/grpcserver/activation_service.go]: ok. in that case do you object that i remove the same check in Get()?

Comment (fasmat) [reply api/grpcserver/activation_service.go]: I wasn't aware that there was such a check, but yes, imo it can go there as well 🙂

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**